[REFACTOR] 카카오 로그인 구현 방식 변경

jye326 opened the pull request:
- 리다이렉트 URL 구성 및 응답 처리 개선
- 기존 회원 로그인 및 신규 회원 회원가입 처리 로직 유지

## Issue Number
closed #1139 

## As-Is

기존의 클라이언트와 카카오 인증 서버가 직접 통신하는 방식은 CSRF 방지를 위한 state 값의 생성이 어려워 보안 상 취약합니다. 이에 카카오가 권장하는 서버 주도적 흐름으로 카카오 로그인을 리팩터링합니다.

## To-Be

### 1. 인증 코드 요청 방법 변경
이제 클라이언트가 직접 카카오 인증 서버로 인증 코드를 요청하지 않습니다. ~`https://kauth.kakao.com/oauth/authorize`~
대신 앱 서버에 `GET kakao/login`를 요청하면 서버가 카카오 인증 서버로 클라이언트를 리다이랙트합니다.

### 2. Redirect URL 변경
클라이언트가 카카오 인증 서버로부터 인증 코드를 받아 서버로 전달해주는 과정이 사라졌습니다.
서버가 직접 인증 코드를 받을 수 있도록 RedirectUrl을 서버 주소로 변경해야 합니다. 
(`www.fittoring.com -> api.fittoring.com`, `dev.fittoring.com -> devapi.fittoring.com`, `localhost:3000 -> localhost:8080`)

### 3. state 값
세션 저장 방식은 SameSite 정책으로 인해 세션 유지가 불가능합니다. 따라서 state값을 JWT로 검증합니다.

### 4. 콜백 함수 리다이랙트 처리
사용자가 카카오 로그인을 성공한 경우 kakao/callback 메서드가 호출됩니다. 이는 클라이언트가 직접 api 요청을 보낸 것이 아니라 카카오 인증 서버가 브라우저를 리다이랙트한 것이므로 kakao/callback의 응답값에 대한 동작을 클라이언트가 직접 할 수 없습니다. 따라서 서버가 적절히 클라이언트를 리다이랙트 해줘야 합니다.
- 환경 별로 리다이랙트 주소가 다르기 때문에 CLIENT_BASE_URL (메인페이지) 환경변수를 추가합니다.
- 로그인 성공시 accessToken과 함께 CLIENT_BASE_URL (메인페이지) 로 리다이랙트합니다.
- 소셜 회원가입 필요시 oauthSignUpToken과 함께 CLIENT_BASE_URL + "/identity-verification" (소셜 회원가입 페이지)로 리다이랙트합니다.


## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description

jye326 committed:
feat: 카카오 로그인 및 콜백 로직 개선

- 카카오 로그인 state 값 생성 및 세션 저장 기능 추가
- state 값 검증 로직 추가로 보안 강화
- 기존 회원 로그인 및 신규 회원 회원가입 처리 로직 수정
- 리다이렉트 URL 구성 및 응답 처리 개선
- 불필요한 import 제거 및 코드 정리

jye326 committed:
feat: 카카오 로그인 및 콜백 로직 개선

- 카카오 로그인 state 값 생성 및 세션 저장 기능 추가
- state 값 검증 로직 추가로 보안 강화
- 기존 회원 로그인 및 신규 회원 회원가입 처리 로직 수정
- 리다이렉트 URL 구성 및 응답 처리 개선
- 불필요한 import 제거 및 코드 정리

jye326 committed:
refactor: AuthService 메서드 명칭 수정으로 일관성 향상

- allowOauthSignIn -> allowOauthLogin, allowOauthSignUp -> allowOauthRegistration으로 메서드 이름 변경
- 코드 명확성과 가독성 개선

jye326 committed:
Revert "refactor: AuthService 메서드 명칭 수정으로 일관성 향상"

This reverts commit 30254d32c917df6b99a2b8b369ffd6cd9284af33.

jye326 committed:
refactor: 컨트롤러 코드 리팩토링 및 Kakao 로그인 리다이렉트 로직 개선

- 불필요한 import 정리 및 wildcard (*) 사용으로 가독성 향상
- Kakao 로그인 리다이렉트 메서드 수정: ResponseEntity로 변환하여 리다이렉트 처리 일관성 유지
- UriComponentsBuilder 활용으로 URL 구성 로직 간소화 및 가독성 개선

jye326 committed:
Merge branch 'refactor/1139-kakao-login' of https://github.com/woowacourse-teams/2025-Fit-toring into refactor/1139-kakao-login

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -99,43 +113,73 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
-    @GetMapping("/kakao/callback")
-    public ResponseEntity<Void> kakaoCallBackSuccess() {
-        return ResponseEntity.status(HttpStatus.OK).build();
+    @GetMapping("kakao/login")
```
p1: 맨 앞에 `/`가 빠졌어요!

jye326 replied:
어머나

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -99,43 +113,73 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
+                .toUri();
+
+        // redirect
+        return ResponseEntity.status(HttpStatus.FOUND).location(url).build();
```
p3: `FOUND` 응답 코드와 `Location` 헤더 덕분에 응답을 받은 클라이언트에서 해당 `url`로 리다이렉트 되는 거군요.

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -99,43 +113,73 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
-    @PostMapping("/kakao/callback")
-    public ResponseEntity<?> kakaoCallback(
+    @GetMapping("/kakao/callback")
+    public ResponseEntity<?> kakaoCallBack_V2(
```
p2: `kakaoCallback()` 메서드가 사라졌는데 `V2`라는 버전을 붙일 필요가 있을지 의문입니다! V1이 없는데 V2만 있는 게 부자연스러운 것 같아서요!

jye326 replied:
기존 메서드를 그대로 남겨두고 버저닝 하려다가 삭제했는데 V2를 지우는 걸 깜빡했네요! 감사합니다~

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -99,43 +113,73 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
+        if (savedState == null || !savedState.equals(state)) {
+            throw new IllegalStateException("로그인 세션 불일치 : state 값이 일치하지 않습니다.");
+        }
+    }
```
p3: 서버와 클라이언트 간 상태 키 검증이 들어갔군요. 👍

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -32,17 +38,25 @@
 public class AuthController {
 
     private static final String REFRESH_TOKEN_COOKIE_NAME = "refreshToken";
+    public static final String KAKAO_STATE = "KAKAO_STATE";
```
p1: 상태 키가 매번 똑같아도 괜찮을까요? 상태 키는 공격자가 예측하지 못하는 값이어야 하는데, 이게 동일하다면 공격자 입장에서 아래와 같이 CSRF 공격이 가능할 것 같아요.
1. 정상적으로 카카오 로그인을 수행해서 `KAKAO_STATE` 값을 알아낸다.
2. 이후 공격 시 동일한 `KAKAO_STATE`를 사용하여 CSRF 공격을 한다.

그래서 개인적으로는 상수보다 랜덤 변수로 처리하는 게 좋을 것 같습니다!

jye326 replied:
STATE값은 `redirectKakaoAuth()`에서 랜덤 변수로 생성하고 있습니다. 해당 상수는 세션 키 값이에요!

kimdevv replied:
헉 그렇네요!

dye0p replied:
p1: KAKAO_STATE는 사용하지 않는 것 같은데, 두신 이유가 있을까요?

전체적으로 사용하지 않는 매개변수나 변수들이 있는 것 같아요. 의도하신걸까요?

jye326 replied:
state값 검증을 세션 방식에서 토큰 방식으로 변경하면서 기존에 사용하던 코드들이 정리가 안됐네요! 꼼꼼한 리뷰 감사합니다!

kimdevv reviewed (changes requested):
고생하셨습니다~ 리뷰하면서 카카오 로그인 방식을 다시 한 번 돌아볼 수 있었던 것 같아요!
몇 가지 의문 사항 커멘트로 남겨 놨습니다! 한 번 확인해 주세요~

jye326 committed:
merge

jye326 committed:
refactor: AuthController 카카오 로그인 및 콜백 메서드 URL 및 명칭 수정

- 카카오 로그인 리다이렉트 URL에 누락된 슬래시 추가 (`kakao/login` → `/kakao/login`)
- 콜백 메서드 명칭 수정 (`kakaoCallBack_V2` → `kakaoCallBack`)으로 코드 명확성 향상

jye326 committed:
refactor: docker-compose.yml 포트 매핑 수정

- MySQL 및 다른 서비스의 포트 매핑 오류 수정 (13306 → 3306, 13307 → 3307)
- 로컬 환경과의 일관성 유지 및 설정 명확성 향상

kimdevv reviewed (approved):
고생하셨습니다. 저는 이만 승인합니다~

jye326 committed:
feat: response_type 추가

jye326 committed:
feat: 카카오 로그인 리다이렉트 URL 처리 로직 수정

- 로그인 성공 시 리다이렉트 URL 응답 추가 (`https://localhost:3000/`)
- 신규 회원 카카오 회원가입 토큰 응답 시 리다이렉트 URL 처리 개선
- 주석 제거 및 불필요한 코드 정리

jye326 committed:
feat: 카카오 로그인 state 토큰 도입 및 리다이렉트 로직 개선

- state 토큰 생성 및 유효성 검증 로직 추가
- 회원가입 및 메인 페이지 리다이렉트 URL 변경
- 불필요한 주석 및 코드 제거
- `application.yml`에 client base-url 설정 추가

jye326 committed:
feat: application-test.yml 설정 업데이트

- CLIENT_BASE_URL 환경 변수를 활용한 client base-url 설정 추가

jye326 committed:
feat: application-test.yml client base-url 로컬 URL로 변경

- client base-url 값을 `${CLIENT_BASE_URL}`에서 `http://localhost:3000`으로 변경하여 테스트 환경 적용

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -89,37 +117,62 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
+                .toUri();
+
+        // redirect
+        return ResponseEntity.status(HttpStatus.FOUND).location(url).build();
```
p3: 상태코드 303 또는 301도 동일하게 리디렉션 되는걸로 아는데 302로 정하신 이유가 궁금합니다~

jye326 replied:
헷갈려서 찾아보고 정리했습니다!

`301 Moved Permanently` : 리소스를 영구적으로 이동할 때 사용합니다. 사용 예시로는 웹사이트 주소를 이전하거나 http -> https 로 변경하는 경우가 있습니다.

`302 Found` : 리소스를 임시로 이동할 때 사용합니다. 브라우저의 경우 거의 항상 요청이 POST -> GET으로 전환하여 이동합니다. 주로 로그인 리다이랙트, OAuth autorize -> provider 이동, callback 처리 후 프론트로 이동 등의 상황에서 사용됩니다. HTTP 스펙 상으로는 모호한 의미지만 브라우저 표준처럼 굳어져 OAuth 예제의 대부분이 302를 사용한다고 합니다. 카카오 인증 서버도 우리 서버로 302 리다이랙트 해줍니다.

`303 See Other` : "요청은 처리됐으니, 결과는 다른 URL에서 GET으로 봐라" 라는 의미입니다. POST -> GET 전환을 명시적으로 보장합니다. 폼 제출 -> 완료 페이지, 로그인 콜백 후 메인 페이지 이동 시 사용됩니다. 의미적으로는 가장 명확한 상태 코드입니다.

스펙적으로 가장 적합한 코드는 303이지만 302도 관행적으로 GET 리다이랙트가 이루어져 OAuth에서 가장 널리 사용되므로 유지하겠습니다.

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -89,37 +117,62 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
                 .build();
     }
 
-    @GetMapping("/kakao/callback")
-    public ResponseEntity<Void> kakaoCallBackSuccess() {
-        return ResponseEntity.status(HttpStatus.OK).build();
+    @GetMapping("/kakao/login")
+    public ResponseEntity<Void> redirectKakaoAuth(HttpServletRequest request, HttpServletResponse response)
+            throws IOException {
+        // SameSite 정책으로 세션 유지 X -> JWT로 state 변경
+        String state = jwtProvider.createStateToken();
+
+        // redirect url 구성
+        URI url = UriComponentsBuilder.fromUriString("https://kauth.kakao.com/oauth/authorize")
+                .queryParam("response_type", "code")
+                .queryParam("client_id", kakaoClientId)
+                .queryParam("redirect_uri", kakaoRedirectUrl)
+                .queryParam("state", URLEncoder.encode(state, StandardCharsets.UTF_8))
+                .build()
+                .toUri();
+
+        // redirect
+        return ResponseEntity.status(HttpStatus.FOUND).location(url).build();
     }
 
-    @PostMapping("/kakao/callback")
-    public ResponseEntity<?> kakaoCallback(
+    @GetMapping("/kakao/callback")
+    public ResponseEntity<Void> kakaoCallBack(
```
p3: 기존의 Post에서 왜 Get으로 바뀐건가요?

jye326 replied:
기존에는 클라이언트가 직접 POST 요청을 보내줬지만, 이제는 카카오 인증서버의 302 콜백을 서버가 직접 받기 때문입니다. 302 콜백은 관행적으로 GET으로 들어옵니다.

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -89,37 +117,62 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
-    public ResponseEntity<Void> kakaoCallBackSuccess() {
-        return ResponseEntity.status(HttpStatus.OK).build();
+    @GetMapping("/kakao/login")
+    public ResponseEntity<Void> redirectKakaoAuth(HttpServletRequest request, HttpServletResponse response)
```
p1: 매개변수 request, response는 사용하지 않는 것 같은데, 놔두신 이유가 있나요?

jye326 replied:
이것도 기존 방식의 잔재네요! 제거하였습니다~

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -89,37 +117,62 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
-        return ResponseEntity.status(HttpStatus.OK).build();
+    @GetMapping("/kakao/login")
+    public ResponseEntity<Void> redirectKakaoAuth(HttpServletRequest request, HttpServletResponse response)
+            throws IOException {
```
p1: 예외도 던지는 부분이 보이지 않네요~

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -89,37 +117,62 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
-    public ResponseEntity<?> kakaoCallback(
+    @GetMapping("/kakao/callback")
+    public ResponseEntity<Void> kakaoCallBack(
             @RequestParam String code,
             @RequestParam(required = false) String error,
             @RequestParam(required = false, value = "error_description") String errorDescription,
             @RequestParam(required = false) String state,
```
p2: 이 값들은 필수로 받아야 하는 값들인가요? 쓰이지 않는 값들도 보여서 그렇습니다!

jye326 replied:
콜백 리다이랙트시 카카오에서 내려주는 error 관련 값들입니다! 리팩터링하면서 사용 부분이 지워졌네요😥 복구해뒀습니다~!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -89,37 +117,62 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
             @RequestParam(required = false, value = "error_description") String errorDescription,
             @RequestParam(required = false) String state,
-            HttpServletResponse httpResponse
+            HttpServletRequest request,
```
p1: request를 사용하는 부분이 안보여요!

jye326 replied:
이것도 남아있었네요~ 지웠습니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -89,37 +117,62 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
+        // 로그인
         LoginInfoDto loginInfoDto = authService.kakaoLogin(code);
         AuthTokenDto authTokenDto = loginInfoDto.authTokenDto();
         LoginResponse loginResponse = new LoginResponse(loginInfoDto.memberId());
```
p1: 이것도 사용하지 않는 변수같아요!

jye326 replied:
지웠습니다~

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/presentation/AuthController.java:
```diff
@@ -89,37 +117,62 @@ public ResponseEntity<Void> verifyPhoneNumber(@RequestBody @Valid VerificationCo
+    public ResponseEntity<Void> redirectKakaoAuth(HttpServletRequest request, HttpServletResponse response)
+            throws IOException {
+        // SameSite 정책으로 세션 유지 X -> JWT로 state 변경
+        String state = jwtProvider.createStateToken();
```
p2: 세션 저장 방식이 불가능한 이유가 이해가 잘 안되네요..ㅠ 설명 부탁드려도 될까요!

jye326 replied:
기존 세션 기반 플로우는 아래와 같습니다.
1. `/kakao/login` 에서 서버가 세션을 만들고 (JSESSIONID 발급) 세션에 `state`값을 저장
2. 브라우저가 이후 `/kakao/callback` 요청 때 같은 JSESSIONID 쿠키를 보내야 세션을 찾아서 `state`를 꺼낼 수 있음

OAuth 콜백은 cross-site redirect라 쿠키가 빠질 수 있습니다.
1. 브라우저 -> 우리 서버 `/kakao/login` 요청
2. 우리서버 -> 카카오로 브라우저를 302 Redirect
3. 카카오 -> 우리서버로 브라우저를 302 Redirect `/kakao/callback?code=...&state=...`

이때 3단계는 카카오에서 우리 서버로 넘어오는 `cross-site redirect` 이고 이때 브라우저는 `SameSite` 규칙에 따라 쿠키를 보낼지 말지 결정합니다.

세션 방식을 사용했을 당시에 쿠키에 `SameSite` 설정을 별도로 해주지 않았으므로 기본값인 `Lax`로 설정되었을 것인데, 이는 `Secure` / `HTTPS`등 환경에 따라 쿠키가 전달되지 않을 수 있는 불안정한 상태입니다. 실제로 HTTP 환경에서 테스트 했을 때는 쿠키가 전송되지 않았습니다.

따라서 환경에 무관하게 안정적인 `state` 검증을 위해 `state`를 `jwt` 토큰으로 변환하여 검증하는 방식으로 변경했습니다!

dye0p commented on backend/fittoring/src/main/java/fittoring/application/auth/service/JwtProvider.java:
```diff
@@ -49,12 +50,18 @@ public String createRefreshToken() {
 
+    public String createStateToken() {
+        Date now = new Date();
+        Date stateMillis = new Date(now.getTime() + Duration.ofMinutes(3).toMillis());
```
p2: 만료시간이 변경될 가능성은 적겠지만, 환경변수로 분리하는 방법도 생각해볼 수 있겠네요!

jye326 replied:
변경될 가능성이 희박할 것 같아 일단 고정해두었습니다~

dye0p commented on frontend/src/pages/login/components/LoginForm/LoginForm.tsx:
```diff
@@ -75,7 +70,7 @@ function LoginForm() {
 
   const handleSocialLoginButtonClick = () => {
-    window.location.href = KAKAO_URL;
+    window.location.href = `${process.env.API_BASE_URL}/kakao/login`
```
클라이언트 코드 수정까지!!👍

jye326 replied:
프론트가 해줬답니다~

dye0p reviewed (commented):
수고하셨어요~ 코멘트 몇가지 남겼습니다! 확인부탁드려요~

jye326 committed:
refactor: AuthController 리팩토링 및 OAuth 콜백 로직 개선

- 불필요한 import 정리 및 wildcard(*) 사용으로 가독성 향상
- Kakao 로그인 리다이렉트 메서드에서 HttpServletRequest 제거 및 리다이렉트 URL 구성 로직 간소화
- OAuth 콜백 메서드에서 오류 처리 로직 추가 및 코드 구조 개선

jye326 reviewed (commented):
꼼꼼하게 확인하고 리뷰 남겨주셔서 감사합니다!! 고칠게 많았네요!

jye326 committed:
merge

jye326 committed:
merge
